### ISSUE TITLE
Hessian not symmetric

### ISSUE THREAD

**Issue by jjough** [tensorflow/tensorflow#10513]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: OS X 10.11.6
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**:  v1.1.0-rc0-61-g1ec6ed5 1.1.0
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**: CPU
- **GPU model and memory**: CPU
- **Exact command to reproduce**:

### Describe the problem
The Hessian should be completely symmetric. Computing the Hessian with tf.gradients gives an almost-symmetric matrix, with errors in symmetry on order of 10^-7; or about 1% of the minimum value of the Hessian.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

```python
import tensorflow as tf
sess = tf.InteractiveSession()

def tf_jacobian(tensor2, tensor1,sess = tf.get_default_session()):
    """
    Computes the tensor d(tensor2)/d(tensor1) recursively.
    :param tensor2: numerator of Jacobian
    :param tensor1: denominator of Jacobian
    :return: a tensor of dimension (dim_tensor2 x dim_tensor1)
    """
    # can't do tensor.get_shape() because shape inference doesn't work for stack/unstack
    shape = list(sess.run(tf.shape(tensor2)))
    if shape:
        # if tensor2 is not a scalar, split tensor2 along first dimension and recur
        # int trick from https://github.com/tensorflow/tensorflow/issues/7754
        tensor2_split = tf.split(axis = 0, num_or_size_splits = int(shape[0]), value = tensor2)
        grad_split = [tf_jacobian(tf.squeeze(M, squeeze_dims = 0), tensor1) for M in tensor2_split]
        return tf.stack(grad_split)
    else:
        # tensor2 is a scalar, so calculate gradient
        grad = tf.gradients(tensor2, tensor1)
        return tf.squeeze(grad, squeeze_dims = [0]) 

def tf_hessian(b, a):
    return tf_jacobian(tf_jacobian(b,a), a)
        
def tf_sym(tensor):
    A = tensor - tf.transpose(tensor, [1,0])
    return tf.reduce_sum(tf.abs(A))

x = tf.constant([1., 2., 3.])
w = tf.random_normal(shape = (3,3))
output = tf.matmul([x],w)
cost = tf.reduce_sum(tf.nn.tanh(output))

hessian = tf_hessian(cost,x)
hessian_sym = tf_sym(hessian)

sess.run(hessian_sym)


>> 2.3841858e-07 # should be exactly zero
```

Edit: simpler example

**Comment by michaelisard** [comment]
@rmlarsen any suggestions?

**Comment by jkschin** [comment]
It almost sounds like non-deterministic reductions. See https://github.com/tensorflow/tensorflow/issues/5527.

**Comment by ekelsen** [comment]
Hmmm...even when running on the CPU, which should be deterministic for these operations, and especially these tiny sizes, it still happens.  I'm not sure if that suggests a bug or that there is a surprising source of non-determinism that this is exposing.

**Comment by jjough** [comment]
Yeah, seems odd, especially as Theano always gives the right answer.

BTW my example was bad because w doesn't have to be a random normal matrix. If w is a small constant, it gives a repeatable, deterministic error:

```python

import tensorflow as tf
sess = tf.InteractiveSession()

def tf_jacobian(tensor2, tensor1,sess = tf.get_default_session()):
    """
    Computes the tensor d(tensor2)/d(tensor1) recursively.
    :param tensor2: numerator of Jacobian
    :param tensor1: denominator of Jacobian
    :return: a tensor of dimension (dim_tensor2 x dim_tensor1)
    """
    # can't do tensor.get_shape() because shape inference doesn't work for stack/unstack
    shape = list(sess.run(tf.shape(tensor2)))
    if shape:
        # if tensor2 is not a scalar, split tensor2 along first dimension and recur
        # int trick from https://github.com/tensorflow/tensorflow/issues/7754
        tensor2_split = tf.split(axis = 0, num_or_size_splits = int(shape[0]), value = tensor2)
        grad_split = [tf_jacobian(tf.squeeze(M, squeeze_dims = 0), tensor1) for M in tensor2_split]
        return tf.stack(grad_split)
    else:
        # tensor2 is a scalar, so calculate gradient
        grad = tf.gradients(tensor2, tensor1)
        return tf.squeeze(grad, squeeze_dims = [0]) 

def tf_hessian(b, a):
    return tf_jacobian(tf_jacobian(b,a), a)
        
def tf_sym(tensor):
    A = tensor - tf.transpose(tensor, [1,0])
    return tf.reduce_sum(tf.abs(A))

x = tf.constant([1., 2., 3.])
w = tf.constant([[ 1.01,  2,  3],
                 [1.02, 2,  3],
                 [1.03, 2,  3]])
output = tf.matmul([x],w)
cost = tf.reduce_sum(tf.nn.tanh(output))

hessian = tf_hessian(cost,x)
hessian_sym = tf_sym(hessian)
print(sess.run(w))
sess.run(hessian_sym)

>> 1.4551915e-11
```

**Comment by jkschin** [comment]
I looked deeper at your code. When we print with more precision, we realize that it is symmetric except one point. Pardon the very long matrix, seeing it in higher precision is better.

[[ -3.7940350011922419071197509765625000000000e-05, -3.8315996789606288075447082519531250000000e-05, **-3.8691643567290157079696655273437500000000e-05**]
 [ -3.8315996789606288075447082519531250000000e-05, -3.8695365219609811902046203613281250000000e-05, -3.9074730011634528636932373046875000000000e-05]
 [ **-3.8691647205268964171409606933593750000000e-05**, -3.9074730011634528636932373046875000000000e-05, -3.9457816455978900194168090820312500000000e-05]]

I've got no ideas at the moment, but I've been running some CPU code recently and I've observed that even tf.nn.conv2d has some weird behaviour. I'm still exploring this and will post the issue once I have some concrete results. Perhaps someone who knows more about machine epsilon can give some hints to this.

**Comment by jjough** [comment]
Thanks, makes a lot more sense.

> When we do some multiplications and additions with these floating point numbers, rounding errors definitely accumulate.

I thought errors cannot accumulate with symbolic computation.

**Comment by ekelsen** [comment]
I talked with @rmlarsen about this and we think the issue that when you compute the two entries that _should_ be the same, they are computed according to a different order of operations (the derivatives are taken in a different order).  Floating point doesn't associate, so you are likely to get different answers.

If you want a symmetric matrix, it would be best to simply compute half of it and then flip it.

**Comment by jjough** [comment]
Thanks, makes sense.

**Comment by jjough** [comment]
Got it, thanks.
